omniscidb/QueryBuilder/QueryBuilder.cpp

```diff
@@ -1621,9 +1662,19 @@ BuilderNode BuilderNode::proj(const ExprPtrVector& exprs,
   return {builder_, proj};
 }
 
-BuilderNode BuilderNode::filter(BuilderExpr condition) const {
+BuilderNode BuilderNode::filter(const BuilderExpr& condition) const {
   checkExprInput(condition, {node_.get()}, "filter");
-  auto filter = std::make_shared<Filter>(condition.expr(), node_);
+  // Filter-out virtual column if it is not used on the filter.,
+  auto base = node_;
+  if (node_->is<Scan>()) {
+    auto used_cols = InputColIndexesCollector::collect(condition.expr());
+    int cols_to_proj =
+        used_cols.count((int)(node_->size() - 1)) ? node_->size() : node_->size() - 1;
```

Comment:
Can `(int)(node_->size() - 1)` be equal to `-1`?

Reply:
It is a scan, so it cannot be empty and its size is at least 2.